community/io/src/test/java/org/neo4j/io/pagecache/PageCacheTest.java

```diff
@@ -2514,6 +2514,24 @@ public void readingAndRetryingOnPageWithOptimisticReadLockingAfterUnmappingMustN
         }
     }
 
+    @Test( timeout = SHORT_TIMEOUT_MILLIS )
+    public void writeCursorShouldNotUnpinSamePage() throws IOException
+    {
+        DefaultPageCacheTracer tracer = new DefaultPageCacheTracer();
+
+        getPageCache( fs, maxPages, pageCachePageSize, tracer );
+
+        try ( PagedFile pagedFile = pageCache.map( file( "a" ), filePageSize );
+              PageCursor cursor = pagedFile.io( 0, PF_SHARED_WRITE_LOCK ) )
+        {
+            assertTrue( cursor.next() );
+            assertTrue( cursor.next( cursor.getCurrentPageId() ) );
+
+            assertThat( "wrong count of pins", tracer.pins(), is( 1L ) );
+            assertThat( "wrong count of unpins", tracer.unpins(), is( 0L ) );
+        }
+    }
```

Comment:
You should also test that an exception is thrown if `cursor.next( cursor.getCurrentPageId() )` is called after the cursor has been closed.